src/typecheck.rs

```diff
@@ -306,6 +304,47 @@ mod tests {
         assert!(typeck_expr(&another_undefined_expr, &ctx).is_err());
     }
 
+    #[test]
+    fn variable_shadowing() {
+        // Test that input variables can't shadow existing variables
+        let program_with_shadowing = r#"
+            alphabet: {'a', 'b'}
+            let c: bool;  // Define 'c' as a program variable
+            on input c {  // Try to use 'c' as an input variable too
+                c = true;
+            }
+            accept if c
+        "#;
+
+        let program: Program = parse(program_with_shadowing).unwrap();
+        let result = typecheck_program(&program);
+
+        // Verify that typechecking fails due to variable shadowing
+        assert!(result.is_err());
```

Comment:
does `assert_eq!(result, TypeError::ShadowedVariable("c"));` not work? It can replace this assert + the match below.